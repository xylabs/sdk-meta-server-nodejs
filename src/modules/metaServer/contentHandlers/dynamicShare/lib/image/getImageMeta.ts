import type { Logger } from '@xylabs/logger'
import { IdLogger } from '@xylabs/logger'
import type {
  Meta, OpenGraphMeta, TwitterMeta,
} from '@xylabs/sdk-meta'
import type { Page } from 'puppeteer'

import { defaultViewportSize, getContentType } from '../../../../lib/index.ts'
import { getImageUrlFromPage } from './getImageUrlFromPage.ts'

/**
 * Generates the expected metadata for the resultant preview image of a Live Share page
 * @param url The url for the page
 * @param width The height of the preview image
 * @param height The width of the preview image
 * @returns The expected metadata for the preview image of a Live Share page
 */
export const getImageMeta = async (
  url: string,
  page: Page,
  width = defaultViewportSize.width,
  height = defaultViewportSize.height,
  logger: Logger = new IdLogger(console, () => `dynamicShare|getImageMeta|${url}`),
): Promise<Meta> => {
  logger.log('generating')
  const imageUrl = await getImageUrlFromPage(url, page)
  if (imageUrl) {
    const type = getContentType(imageUrl) || 'image/png'
    const og: OpenGraphMeta = {
      image: {
        '': imageUrl, height, 'secure_url': imageUrl, type, 'url': imageUrl, width,
      },
    }
    const twitter: TwitterMeta = { card: 'summary_large_image', image: { '': imageUrl } }
    const meta = { og, twitter }
    logger.log(`generated [${imageUrl}]`)
    return meta
  }
  return {}
}
