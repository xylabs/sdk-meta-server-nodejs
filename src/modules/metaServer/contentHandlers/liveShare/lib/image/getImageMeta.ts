import { Meta, OpenGraphMeta, TwitterMeta } from '@xyo-network/sdk-meta'

import { defaultViewportSize } from '../../../../lib/index.js'
import { getImageUrlFromPageUrl } from '../url/index.js'

/**
 * Generates the expected metadata for the resultant preview image of a Live Share page
 * @param url The url for the page
 * @param width The height of the preview image
 * @param height The width of the preview image
 * @returns The expected metadata for the preview image of a Live Share page
 */
export const getImageMeta = (url: string, width = defaultViewportSize.width, height = defaultViewportSize.height): Meta => {
  console.log(`[liveShare][getImageMeta][${url}]: generating`)
  const imageUrl = getImageUrlFromPageUrl(url, width, height)
  const og: OpenGraphMeta = { image: { '': imageUrl, height, 'secure_url': imageUrl, 'type': 'image/png', 'url': imageUrl, width } }
  const twitter: TwitterMeta = { card: 'summary_large_image', image: { '': imageUrl } }
  const meta = { og, twitter }
  console.log(`[liveShare][getImageMeta][${url}]: generated`)
  return meta
}
