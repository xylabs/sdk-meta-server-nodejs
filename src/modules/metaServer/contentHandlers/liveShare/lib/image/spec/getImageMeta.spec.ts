import { getImageMeta } from '../getImageMeta.ts'

describe('getImageMeta', () => {
  const cases = [
    'http://xyo.network',
    'https://beta.foreventory.com/web3/nfts/wallet/0xacdaEEb57ff6886fC8e203B9Dd4C2b241DF89b7a/score/e1b5b2a04069019a2ed2eb14e983b1dfaff96769a98c84fb9502d89615e07398?chainId=1',
  ]
  it.each(cases)('generates page preview image Meta', (url) => {
    const actual = getImageMeta(url)
    expect(actual).toMatchSnapshot()
  })
})
