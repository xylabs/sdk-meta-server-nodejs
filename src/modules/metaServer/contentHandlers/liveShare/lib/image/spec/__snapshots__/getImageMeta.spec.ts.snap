// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getImageMeta > generates page preview image Meta 1`] = `
{
  "og": {
    "image": {
      "": "http://xyo.network/preview/1200/630/img.png",
      "height": 630,
      "secure_url": "http://xyo.network/preview/1200/630/img.png",
      "type": "image/png",
      "url": "http://xyo.network/preview/1200/630/img.png",
      "width": 1200,
    },
  },
  "twitter": {
    "card": "summary_large_image",
    "image": {
      "": "http://xyo.network/preview/1200/630/img.png",
    },
  },
}
`;

exports[`getImageMeta > generates page preview image Meta 2`] = `
{
  "og": {
    "image": {
      "": "https://beta.foreventory.com/web3/nfts/wallet/0xacdaEEb57ff6886fC8e203B9Dd4C2b241DF89b7a/score/e1b5b2a04069019a2ed2eb14e983b1dfaff96769a98c84fb9502d89615e07398/preview/1200/630/img.png?chainId=1",
      "height": 630,
      "secure_url": "https://beta.foreventory.com/web3/nfts/wallet/0xacdaEEb57ff6886fC8e203B9Dd4C2b241DF89b7a/score/e1b5b2a04069019a2ed2eb14e983b1dfaff96769a98c84fb9502d89615e07398/preview/1200/630/img.png?chainId=1",
      "type": "image/png",
      "url": "https://beta.foreventory.com/web3/nfts/wallet/0xacdaEEb57ff6886fC8e203B9Dd4C2b241DF89b7a/score/e1b5b2a04069019a2ed2eb14e983b1dfaff96769a98c84fb9502d89615e07398/preview/1200/630/img.png?chainId=1",
      "width": 1200,
    },
  },
  "twitter": {
    "card": "summary_large_image",
    "image": {
      "": "https://beta.foreventory.com/web3/nfts/wallet/0xacdaEEb57ff6886fC8e203B9Dd4C2b241DF89b7a/score/e1b5b2a04069019a2ed2eb14e983b1dfaff96769a98c84fb9502d89615e07398/preview/1200/630/img.png?chainId=1",
    },
  },
}
`;
